lib/tbot/service_ssh_multiplexer.go

```diff
@@ -0,0 +1,709 @@
+	ctx, span := tracer.Start(
+		ctx,
+		"SSHMultiplexerService/handleConn",
+		oteltrace.WithNewRoot(),
+	)
+	defer func() { tracing.EndSpan(span, err) }()
+	defer downstream.Close()
+
+	// The first thing downstream will send is the multiplexing request which is
+	// the "[host]:[port]\x00" format.
+	buf := bufio.NewReader(downstream)
+	reqBytes, err := buf.ReadBytes('\x00')
+	if err != nil {
+		return trace.Wrap(err, "reading request")
+	}
+	reqBytes = reqBytes[:len(reqBytes)-1] // Drop the NULL.
+	splits := bytes.Split(reqBytes, []byte{':'})
+	if len(splits) != 2 {
+		return trace.BadParameter("malformed request")
+	}
+	host := string(splits[0])
+	port := string(splits[1])
+
+	log := s.log.With(
+		slog.Group("req",
+			"host", host,
+			"port", port,
+		),
+	)
+	log.InfoContext(ctx, "Received multiplexing request")
+
+	clusterName := s.identity.Get().ClusterName
+	expanded, matched := tshConfig.ProxyTemplates.Apply(
+		net.JoinHostPort(host, port),
+	)
+	if matched {
+		log.DebugContext(
+			ctx,
+			"Proxy templated matched",
+			"populated_template", expanded,
+		)
+		if expanded.Cluster != "" {
+			clusterName = expanded.Cluster
+		}
+
+		if expanded.Host != "" {
+			host = expanded.Host
+		}
+	}
+
+	var target string
+	if expanded == nil || (len(expanded.Search) == 0 && expanded.Query == "") {
+		host = cleanTargetHost(host, proxyHost, clusterName)
+		target = net.JoinHostPort(host, port)
+	} else {
+		node, err := resolveTargetHostWithClient(ctx, authClient, expanded.Search, expanded.Query)
+		if err != nil {
+			return trace.Wrap(err, "resolving target host")
+		}
+
+		log.DebugContext(
+			ctx,
+			"Found matching SSH host",
+			"host_uuid", node.GetName(),
+			"host_name", node.GetHostname(),
+		)
+
+		target = net.JoinHostPort(node.GetName(), "0")
+	}
+
+	upstream, _, err := hostDialer.DialHost(ctx, target, clusterName, nil)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if s.cfg.SessionResumptionEnabled() {
+		log.DebugContext(ctx, "Enabling session resumption")
+		upstream, err = resumption.WrapSSHClientConn(
+			ctx,
+			upstream,
+			func(ctx context.Context, hostID string) (net.Conn, error) {
+				log.DebugContext(ctx, "Resuming connection")
+				// if the connection is being resumed, it means that
+				// we didn't need the agent in the first place
+				var noAgent agent.ExtendedAgent
+				conn, _, err := hostDialer.DialHost(
+					ctx, net.JoinHostPort(hostID, "0"), clusterName, noAgent,
+				)
+				return conn, err
+			})
+		if err != nil {
+			return trace.Wrap(err, "wrapping conn for session resumption")
+		}
+	}
+
+	// Drain the buffer we used to read in the request in case it read in more
+	// than just the initial request.
+	if _, err := io.CopyN(upstream, buf, int64(buf.Buffered())); err != nil {
+		err := trace.Wrap(err, "draining request buffer to upstream")
+		if err := upstream.Close(); err != nil {
+			return trace.NewAggregate(
+				trace.Wrap(err, "closing upstream"), err,
+			)
+		}
+		return err
+	}
```

Comment:
Pretty sure that this has the potential to deadlock, since this happens before the call to ProxyConn starts moving data in the opposite direction. Tying the data forwarding (through `utils.ProxyConn`) to `egCtx` also results in all connections being terminated if tbot fails to renew credentials, which is kind of weird - was that intended? Is that because we should really just exit and try spawning tbot again, or what?

Reply:
> also results in all connections being terminated if tbot fails to renew credentials, which is kind of weird - was that intended? Is that because we should really just exit and try spawning tbot again, or what?

Indeed - if this is in an unhealthy state, we need to recover by potentially restarting. Realistically, even if I extend the number of retries on the credential renewal here, other components of `tbot` will become unhealthy and lead to a restart anyway. I think I'd like to do another PR soon that has a longer retry period/better unavailability behaviour but it's a bit beyond the scope o fthis.

Reply:
[8f574a9](https://github.com/gravitational/teleport/pull/42592/commits/8f574a95f70c3a78cd439d0d42089b59e3ee67f4)